src/game/client/draw/w3dmodeldraw.cpp

```diff
@@ -1154,3 +1179,1913 @@ void W3DModelDrawModuleData::Parse_Condition_State(INI *ini, void *instance, voi
+                }
+            }
+        }
+
+        if (m_renderObject != nullptr) {
+            if (m_curState != nullptr) {
+                if (m_whichAnimInCurState != -1) {
+                    if (m_curState->m_animations[m_whichAnimInCurState].Is_Idle_Anim()
+                        || Test_Animation_Flag(m_curState->m_flags, RESTART_ANIM_WHEN_COMPLETE)) {
+                        Adjust_Animation(m_curState, -1.0f);
+                    }
+                }
+            }
+        }
+    }
+
+    Adjust_Anim_Speed_To_Movement_Speed();
+
+    if (m_renderObject != nullptr) {
+        Matrix3D m = *transform;
+        Adjust_Transform_Mtx(m);
+        m_renderObject->Set_Transform(m);
+    }
+
+    Handle_Client_Turret_Positioning();
+    Recalc_Bones_For_Client_Particle_Systems();
+
+    if (Get_W3D_Model_Draw_Module_Data()->m_particlesAttachedToAnimatedBones) {
+        Update_Bones_For_Client_Particle_Systems();
+    }
+
+    Handle_Client_Recoil();
+}
+
+const ModelConditionInfo *W3DModelDraw::Find_Transition_For_Sig(uint64_t sig) const
+{
+    auto trans = Get_W3D_Model_Draw_Module_Data()->m_transitionMap.find(sig);
+
+    if (trans != Get_W3D_Model_Draw_Module_Data()->m_transitionMap.end()) {
+        return &trans->second;
+    } else {
+        return nullptr;
+    }
+}
+
+bool Maintain_Frame_Across_States(int flags)
+{
+    return (flags & 0x3A0) != 0;
+}
+
+bool Maintain_Frame_Across_States_2(int flags, int flags2)
+{
+    return (flags2 & 0x3A0 & flags & 0x3A0) != 0;
+}
+
+float W3DModelDraw::Get_Current_Anim_Fraction() const
+{
+    if (m_curState == nullptr || !Maintain_Frame_Across_States(m_curState->m_flags) || !m_renderObject
+        || m_renderObject->Class_ID() != RenderObjClass::CLASSID_HLOD) {
+        return -1.0f;
+    }
+
+    HLodClass *hlod = static_cast<HLodClass *>(m_renderObject);
+
+    float frame;
+    int frames;
+    int mode;
+    float multiplier;
+    hlod->Peek_Animation_And_Info(frame, frames, mode, multiplier);
+
+    if (frame < 0.0f) {
+        return 0.0f;
+    }
+
+    if (frames > frame) {
+        return frame / (frames - 1.0f);
+    }
+
+    return 1.0f;
+}
+
+void W3DModelDraw::Adjust_Animation(ModelConditionInfo const *prev_state, float f)
+{
+    if (m_curState != nullptr) {
+        int count = m_curState->m_animations.size();
+
+        if (count <= 0) {
+            m_whichAnimInCurState = -1;
+        } else {
+            if (count == 1) {
+                m_whichAnimInCurState = 0;
+            } else if (prev_state == m_curState) {
+                int which = m_whichAnimInCurState;
+
+                while (m_whichAnimInCurState == which) {
+                    m_whichAnimInCurState = Get_Client_Random_Value(0, count - 1);
+                }
+            } else {
+                m_whichAnimInCurState = Get_Client_Random_Value(0, count - 1);
+            }
+
+            HAnimClass *anim = m_curState->m_animations[m_whichAnimInCurState].Get_Anim_Handle();

```

Comment:
Please add a TODO comment here that this needs checking. anim can eventually leak here.

Reply:
fixed.